app/models/miq_filter.rb

```diff
@@ -43,28 +43,43 @@ def self.belongsto2object_list(tag)
     # /belongsto/ExtManagementSystem|<name>/EmsCluster|<name>/EmsFolder|<name>
     raise _("invalid tag: %{tag}") % {:tag => tag} unless tag.starts_with?("/belongsto/ExtManagementSystem")
     parts = tag.split("/")[2..-1]
+    depth = parts.size - 1 # ancestry uses 0 based depth
 
-    # root object
-    klass, name = parts.shift.split("|")
-    obj = find_object_by_name(klass, name)
+    # Get the root EMS object
+    # TODO: For tree queries deeper than 1, we don't actually need the ems object,
+    #       so find a way to just get the id
+    ems_class, ems_name = parts.first.split("|", 2)
+    ems = find_object_by_name(ems_class, ems_name)
 
-    if obj.nil?
-      _log.warn("lookup for klass=#{klass.to_s.inspect} with name=#{name.inspect} failed in tag=#{tag.inspect}")
+    if ems.nil?
+      _log.warn("lookup for klass=#{ems_class.to_s.inspect} with name=#{ems_name.inspect} failed in tag=#{tag.inspect}")
       return []
     end
 
-    # traverse the tree
-    parts.each_with_object([obj]) do |p, result|
-      tag_part_klass, name = p.split("|")
-      tag_part_klass = tag_part_klass.constantize
+    return [ems] if depth == 0
 
-      obj = obj.with_relationship_type('ems_metadata') do
-        obj.children.grep(tag_part_klass).detect { |c| c.name == name }
-      end
+    # Get the leaf node object for this EMS
+    leaf_class, leaf_name = parts.last.split("|", 2)
+    leaves = leaf_class.constantize
+      .includes(:all_relationships)
+      .where(:name => leaf_name, :ems_id => ems.id)
+
+    # If multiple leaves come back, filter by depth, and then find which one has
+    #   the valid path. It's possible multiple leaves could be at the same depth.
+    leaves.each do |leaf|
+      next unless leaf.depth == depth
```

Comment:
I guess the problem would depend on which object the ancestor_ids is coming from.  If it's getting the _real_ ancestor ids (i.e. the resource_ids), or is it just the relationship record ids.  If the latter, it should not incur a penalty because relationship ancestor_ids (I think) is basically just an in-memory string split of the ancestry column (i.e. no queries)

Reply:
I double checked and the depth method in ancestry gem is `ancestor_ids.size`, as you said, and `ancestor_ids` is just parsing the ancestry column which should already be in memory.

[`RelationshipMixin#depth`](https://github.com/ManageIQ/manageiq/blob/9374d98b8b62e2aedad2d207030a70d194d79768/app/models/mixins/relationship_mixin.rb#L475-L476) pulls that value out of the resource's relationship, which has already been queried, so this will not introduce extra queries.

(Also, anecdotally, I don't see any extra queries when i manually inspect them in the log)

Reply:
thnx for looking into this.